Fix Android env var collection

cmannett85-arm opened the pull request:
RD uses set/getprop to emulate env vars configured from the host, and it uses Process::LaunchProcess to do this that in turn uses fork/execve to launch getprop.  execve doesn't take the shell lookup patterns to find getprop so it ultimately uses FileIO::FindFileInPath(..) to prepend the appropriate path - but this isn't implemented in the Android version.  This resulted in the capture options using default values.

This fix is just an almost-copy of the default POSIX implementation.  I also added slightly more error output to the execve failure logging too.

baldurk commented:
Is the only issue here that we're using `execve` instead of `execvpe` so android isn't searching PATH for us? If that's the problem it seems more reasonable to use `execvpe` rather than trying to manually replicate that behaviour ourselves.

cmannett85-arm commented:
In principle yes, but I was very wary of making a such a low-level change across all POSIX platforms - some of which I'm unable to test on.  There's nothing wrong with the general POSIX code now, the issue I saw was because a small part of it wasn't implemented for Android, so I just fixed that bit.

baldurk commented:
Not searching PATH was definitely never a deliberate behaviour so I'd rather do that - it's how the windows implementation of `LaunchProcess` behaves and it's how the code expects it to work as we wouldn't deliberately have that difference between platforms. I'm surprised this isn't broken elsewhere since e.g. running adb or android commands on linux goes through this path and it does not pass absolute paths to executables.

The change as-is in theory is fine but in practice makes me really worried about doing anything non-trivial and low level like this on Android because you never know when it might break on a random device.

FWIW the only supported POSIX platforms are linux and Android. Mac exists but is not supported so if it's broken that's not a problem, GGP was recently deleted.

cmannett85-arm commented:
> I'm surprised this isn't broken elsewhere since e.g. running adb or android commands on linux goes through this path and it does not pass absolute paths to executables.

It looks like the full path for ADB and others is calculated, see `android_tools.cpp`.  But none of this is an issue on the host as the default POSIX PATH lookup code is used.

> FWIW the only supported POSIX platforms are linux and Android.

Fair enough, I see separate BSD code sections which I'm set up for, but if it's not officially supported.

baldurk commented:
Actually yeh I guess adb isn't a great example because we first try and search through the labyrinthine mess of SDKs and alternatives to find it because it's not generally expected to be available in PATH. If the SDK is found we do try and execute it through PATH but that won't be hit as often (and so maybe doesn't work?).

Either way searching PATH via execvpe should be a no-op when passing an absolute path, and will only affect cases that don't work with execve. I'm not really sure the point of execve except perhaps as a bullet-proofing security thing where you want to avoid overriding via PATH but that makes more sense as an opt-out rather than opt-in. There is certainly the argument that there might be android systems where execvpe doesn't work of course, but that's just the reality of android where everything can and will break somewhere.

cmannett85-arm commented:
So it turns out `execvpe` is a GNU thing and Mac doesn't support it.

baldurk commented:
Computers were mistake.

For now since mac isn't supported do something like

```
#if ENABLED(RDOC_APPLE)
#define execvpe execve
#endif
```

and it can be fixed properly later for mac as needed.

baldurk commented on renderdoc/os/posix/posix_process.cpp:
```diff
@@ -626,8 +628,8 @@ static pid_t RunProcess(rdcstr appName, rdcstr workDir, const rdcstr &cmdLine, c
       chdir(workDir.c_str());
-      execve(appPath.c_str(), argv, envp);
-      fprintf(stderr, "exec failed\n");
+      RD_EXECVE(appPath.c_str(), argv, envp);
```
extreme nit: can you change this to `RD_EXECVPE` since that's what we want to use, it's just not available on OSX.

cmannett85-arm replied:
Done.

cmannett85-arm committed:
Fix Android env var collection

RD uses getprop to emulate geting env vars configured from the host, and it uses Process::LaunchProcess to do this that in turn uses fork/execve to launch getprop.  But execve doesn't use the env's shell lookup mechanism for exes - this resulted in the getprop call failing and the capture options using default values.

This fix is to switch to using execvpe.
